[FEATURE] Argument parser version check (#955)
seqan/seqan3: doc/tutorial/argument_parser/index.md

[APP INFO] :: If this app is developed by SeqAn, visit www.seqan.de for updates.
[APP INFO] :: If you don't want to receive this message again set --version_check 0
```

Case 2: `NDEBUG` is NOT set.
If an application is built in debug mode, we will assume you are the app developer.
Therefore we will inform you whenever a new library version is available:

```
[SEQAN INFO] :: There is a newer SeqAn version available!
[SEQAN INFO] :: Please visit www.seqan.de for an update or inform the developer of this app.
[SEQAN INFO] :: If you don't want to receive this message again set --version-check 0
```
If the application is not registered with the server, you will be prompted to register it.
This will make the distribution of your application version simple and convenient.

```
[SEQAN INFO] :: Thank you for using SeqAn!
[SEQAN INFO] :: You might want to register you app for support and version check features?
[SEQAN INFO] :: Just send us an email to support@seqan.de with your app name and version number.
[SEQAN INFO] :: If you don't want to receive this message any more set --version_check 0
```

### We value your privacy!

> There is no form of user identification, no tracking and we will never give away any collected data to third parties.

Reviewer: ```suggestion
The version check is implemented as a web-request that involves sending certain necessary information.
```

See our [wiki entry](https://github.com/seqan/seqan3/wiki/Update-Notifications) for detailed information.

### Configuring this feature

> Developers that wish to disable this feature permanently you may build the application

Reviewer: ```suggestion
Developers that wish to disable this feature permanently can pass an extra constructor argument:
```

```
[SEQAN INFO] :: Thank you for using SeqAn!
[SEQAN INFO] :: You might want to register you app for support and version check features?
[SEQAN INFO] :: Just send us an email to support@seqan.de with your app name and version number.
[SEQAN INFO] :: If you don't want to receive this message any more set --version_check 0
```

### We value your privacy!

The version check is implemented as a web-request that involves sending certain necessary information.
See our [wiki entry](https://github.com/seqan/seqan3/wiki/Update-Notifications) for detailed information.

### Configuring this feature

Developers that wish to disable this feature permanently can pass an extra constructor argument:
with an extra constructor argument:

\include doc/tutorial/argument_parser/disable_version_check.cpp

Users of applications that have this feature activated can opt-out, by either:

 * disabling it for a specific application simply by setting the option `--version-check 0` or
 * disabling it for all applications by setting the `SEQAN3_NO_VERSION_CHECK` environment variable.

Note that in case there is no `--version-check` option (display available options with `-h/--help)`,

> then the developer already disabled the version check functionality.

Author: that's the one you gave wrote above right? then it's done.

\solution
\snippet doc/tutorial/argument_parser/solution6.cpp file_validator
\endsolution

> 

Reviewer: double newline

> # The Version Update Feature

Reviewer: I would call it "Update Notifications" since we are not really offering automatic updates.

SeqAn applications that are using the seqan3::argument_parser can check SeqAn servers for version updates.
The functionality helps getting new versions out to users faster.
It is also used to inform application developers of new versions of the SeqAn library
which means that applications ship with less bugs.
For privacy implications, please see: https://github.com/seqan/seqan3/wiki/Update-Notifications

### Customized Messages

The version information you receive depends on whether you are an application user or developer.
We differentiate this by inquiring the `NDEBUG` (no-debug) macro.

Case 1: `NDEBUG` is set.

> It's the default in our application and represents that you are a user.

Reviewer: ```suggestion
This is the default for all released binaries so we assume you are an application user.
```

The only message you will eventually encounter is the following:

```
[APP INFO] :: There is a newer version of this application available.

> [APP INFO] :: If this app is developed by SeqAn, visit www.seqan.de for updates.

Reviewer: We should definetely set a user provided URL here.

Reviewer: Compromise: Can we add a `update_url` string member that gets printed here? Any tag-checking can be implemented later, but having that would already help greatly here. It's also something we can tell the developer to set in debug mode. And later, if we recognise the url-type, we won't tell people to register at all.

[APP INFO] :: If you don't want to receive this message again set --version_check 0
```

Case 2: `NDEBUG` is NOT set.

> If you build one of our application or your own one in debug mode, we will consider you as a developer.

Reviewer: ```suggestion
If an application is built in debug mode, we will assume you are the app developer.
```

Therefore we will inform you whenever a new library version is available:

```
[SEQAN INFO] :: There is a newer SeqAn version available!
[SEQAN INFO] :: Please visit www.seqan.de for an update or inform the developer of this app.
[SEQAN INFO] :: If you don't want to recieve this message again set --version-check 0
```
If you are working on your own application, using the SeqAn ArgumentParser,
we will inform you about the possibility to register your application with us.

> This will make the distribution of your application version simple and convenient.

Reviewer: "If the application is not registered with the server, you will be prompted to register it."

### The information we do (not) collect

The process of checking for a new version happens at most once a day
and takes at most three seconds enforced by an internal time-out.

\note The runtime of your application might be slightly affected by the process of checkng the version.
      You might want to temporarily switch off the option while doing sensible performance measurements (--version-check 0).

The following information is transmitted to the servers solely via the URL:

 * The application name and its version
 * The SeqAn version that the application was built with
 * The operating system type (Linux, macOS, Windows or BSD)
 * The CPU type (32 or 64bit)

The purpose of this transmission is to provide accurate update data for your app.
Beyond the update information, we may count the total number of version requests
and may also resolve them to geographical regions.
This may be used for anonymized analysis by the SeqAn team, but raw data is never shared with third parties.

\attention There is no form of user identification and no tracking.
           IP-Addresses are never stored permanently.
           SeqAn collects no information regarding your use of the application,

>            like selected options or arguments provided, and of course no information on your files!

Reviewer: We should move this section to the Wiki. And only link to the wiki here.

> ### Disable this feature any time you want

Reviewer: ```suggestion
### Configuring this feature
```

If you, as a **developer**, still feel uncomfortable with the version check,

> you may build the application with an extra constructor argument that disables the version check permanently:

Reviewer: Developers that wish to set a different notification server, can to the following:

Developers that wish to disable this feature permanently for their application should instead:

Users of applications that have this feature activated can opt-out, by either:
